pkg/operator/staticpod/controller/installerstate/installer_state_controller.go

```diff
@@ -0,0 +1,218 @@
+		timeNowFn:       time.Now,
+	}
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *InstallerStateController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+	}
+}
+
+// degradedConditionNames lists all supported condition types.
+var degradedConditionNames = []string{
+	"InstallerPodPendingDegraded",
+	"InstallerPodContainerWaitingDegraded",
+}
+
+func (c *InstallerStateController) sync() error {
+	pods, err := c.podsGetter.Pods(c.targetNamespace).List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": "installer"}).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// collect all pods that are in pending state for longer than maxToleratedPodPendingDuration
+	pendingPods := []*v1.Pod{}
+	hasPendingPods := false
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != v1.PodPending {
+			continue
+		}
+		// we need to track this so we can requeue faster when there are pending pods
+		hasPendingPods = true
+
+		if pod.Status.StartTime == nil {
+			continue
+		}
+		pendingTime := c.timeNowFn().Sub(pod.Status.StartTime.Time)
+		if pendingTime >= maxToleratedPodPendingDuration {
+			pendingPods = append(pendingPods, pod.DeepCopy())
+		}
+	}
+
+	// in theory, there should never be two installer pods pending as we don't roll new installer pod
+	// until the previous/existing pod has finished its job.
+	foundConditions := c.handlePendingInstallerPods(pendingPods)
+
+	updateConditionFuncs := []v1helpers.UpdateStaticPodStatusFunc{}
+
+	// check the supported degraded foundConditions and check if any pending pod matching them.
+	for _, degradedConditionName := range degradedConditionNames {
+		// clean up existing foundConditions
+		updatedCondition := operatorv1.OperatorCondition{
+			Type:   degradedConditionName,
+			Status: operatorv1.ConditionFalse,
+		}
+		if condition := v1helpers.FindOperatorCondition(foundConditions, degradedConditionName); condition != nil {
+			updatedCondition = *condition
+		}
+		updateConditionFuncs = append(updateConditionFuncs, v1helpers.UpdateStaticPodConditionFn(updatedCondition))
+	}
+
+	if _, _, err := v1helpers.UpdateStaticPodStatus(c.operatorClient, updateConditionFuncs...); err != nil {
+		return err
+	}
+
+	// if we found pending pods, requeue faster and do not wait for informer as the pod can be stuck in Pending
+	// with no update for long time and we want to track that time.
+	if hasPendingPods {
+		c.queue.AddAfter(installerStateControllerWorkQueueKey, 5*time.Second)
```

Comment:
too fast, you'll be thrashing on status changes just updating a time and spewing out events.

Reply:
@deads2k 1 minute?